### PR TITLE
topolgical_sort, lexicographical_topological_sort

### PR DESCRIPTION
**Pull request by MisterSheik**
Rewrote topolgical_sort as a generator:
* It no longer accepts reverse or nbunch arguments
* It is slightly faster.

Added lexicographical_topological_sort, which accepts a key.
Some tests for topological sort were removed because they relied on
CPython's particular dict element ordering.

Added utils from more_itertools:
* These were necessary for working with the new generators.
* "pairwise" is equavalent to zip(some_list, some_list[1:]) -- for
generators.
* "consume" consumes an iterator.

### CODE REVIEW

**Comment by MisterSheik** [comment]
@jtorrents Any chance of a code review?

**Comment by MisterSheik** [comment]
@hagberg  @jfinkels ping

File: networkx/algorithms/dag.py [hunk 1]
```diff
@@ -79,172 +82,184 @@ def is_directed_acyclic_graph(G):
     if not G.is_directed():
         return False
     try:
-        topological_sort(G, reverse=True)
+        consume(topological_sort(G))
```
Comment (jfinkels) [inline networkx/algorithms/dag.py]: Why is it necessary to consume the whole iterator here?

File: networkx/algorithms/dag.py [hunk 2]
```diff
@@ -79,172 +82,184 @@ def is_directed_acyclic_graph(G):
...
         return True
     except nx.NetworkXUnfeasible:
         return False
 
 
-def topological_sort(G, nbunch=None, reverse=False):
```
Comment (jfinkels) [inline networkx/algorithms/dag.py]: I agree that `reverse` should be removed here, but is it possible to leave `nbunch` for backward-compatibility?

File: networkx/algorithms/dag.py [hunk 3]
```diff
@@ -79,172 +82,184 @@ def is_directed_acyclic_graph(G):
...
+    is_directed_acyclic_graph, lexicographical_topological_sort
 
     References
     ----------
-    .. [1] Skiena, S. S. The Algorithm Design Manual  (Springer-Verlag, 1998). 
-        http://www.amazon.com/exec/obidos/ASIN/0387948600/ref=ase_thealgorithmrepo/
+    .. [1] Manber, U. (1989). Introduction to algorithms - a creative approach. Addison-Wesley.
+        http://www.amazon.com/Introduction-Algorithms-A-Creative-Approach/dp/0201120372
     """
     if not G.is_directed():
```
Comment (jfinkels) [inline networkx/algorithms/dag.py]: Use `networkx.utils.not_implemented_for` here.

File: networkx/algorithms/dag.py [hunk 4]
```diff
@@ -79,172 +82,184 @@ def is_directed_acyclic_graph(G):
...
-                fringe.extend(new_nodes)
-            else:           # No new nodes so w is fully explored
-                explored.add(w)
-                order.append(w)
-                fringe.pop()    # done considering this node
-    if reverse:
-        return order
-    else:
-        return list(reversed(order))
+    indegree_map = G.in_degree()
```
Comment (jfinkels) [inline networkx/algorithms/dag.py]: Maybe simpler, to remove key deletion and collapse four lines of code into two:

    indegree = {v: d for v, d in G.in_degree() if d > 0}
    indegree_zero = [v for v, d in G.in_degree() if d == 0]

(Assuming that `in_degree` returns an iterator over pairs, as it will once some current pull requests get merged.)

File: networkx/algorithms/dag.py [hunk 5]
```diff
@@ -79,172 +82,184 @@ def is_directed_acyclic_graph(G):
...
+                     if indegree == 0]
+
+    for node in zero_indegree:
+        del indegree_map[node]
+
+    while zero_indegree:
+        node = zero_indegree.pop()
+        if node not in G:
+            raise RuntimeError("Graph changed during iteration")
+        for _, child in G.edges_iter(node):
```
Comment (jfinkels) [inline networkx/algorithms/dag.py]: Just use `G.edges`, it will become an iterator in the near future.

File: networkx/algorithms/dag.py [hunk 6]
```diff
@@ -79,172 +82,184 @@ def is_directed_acyclic_graph(G):
...
+        raise nx.NetworkXUnfeasible("Graph contains a cycle or graph changed "
+                                    "during iteration")
 
-def topological_sort_recursive(G, nbunch=None, reverse=False):
-    """Return a list of nodes in topological sort order.
 
-    A topological sort is a nonunique permutation of the nodes such
-    that an edge from u to v implies that u appears before v in the
-    topological sort order.
+def lexicographical_topological_sort(G, key=(lambda x: x)):
```
Comment (jfinkels) [inline networkx/algorithms/dag.py]: Should be `key=None` to match the signature of `sorted`, for example. Just check for `None` in the body of the function.

File: networkx/algorithms/dag.py [hunk 7]
```diff
@@ -79,172 +82,184 @@ def is_directed_acyclic_graph(G):
...
     --------
     topological_sort
-    is_directed_acyclic_graph
 
+    References
+    ----------
+    .. [1] Manber, U. (1989). Introduction to algorithms - a creative approach. Addison-Wesley.
+        http://www.amazon.com/Introduction-Algorithms-A-Creative-Approach/dp/0201120372
     """
     if not G.is_directed():
```
Comment (jfinkels) [inline networkx/algorithms/dag.py]: Use `not_implemented_for` here.

File: networkx/algorithms/dag.py [hunk 8]
```diff
@@ -79,172 +82,184 @@ def is_directed_acyclic_graph(G):
...
 
-    def _dfs(v):
-        ancestors.add(v)
+    def create_tuple(node):
+        return key(node), node
 
-        for w in G[v]:
-            if w in ancestors:
-                raise nx.NetworkXUnfeasible("Graph contains a cycle.")
+    indegree_map = G.in_degree()
```
Comment (jfinkels) [inline networkx/algorithms/dag.py]: Same suggestion as above.

File: networkx/algorithms/dag.py (hunk 3 above)
Comment (MisterSheik) [inline networkx/algorithms/dag.py]: Okay.  I considered that, but it will raise a different exception FYI.

File: networkx/algorithms/dag.py (hunk 4 above)
Comment (MisterSheik) [inline networkx/algorithms/dag.py]: Good idea.

File: networkx/algorithms/dag.py (hunk 5 above)
Comment (MisterSheik) [inline networkx/algorithms/dag.py]: Ok.

File: networkx/algorithms/dag.py [hunk 9]
```diff
@@ -79,172 +82,184 @@ def is_directed_acyclic_graph(G):
...
-        topological_sort(G, reverse=True)
+        consume(topological_sort(G))
         return True
     except nx.NetworkXUnfeasible:
         return False
 
 
-def topological_sort(G, nbunch=None, reverse=False):
-    """Return a list of nodes in topological sort order.
+def topological_sort(G):
```
Comment (jfinkels) [inline networkx/algorithms/dag.py]: Needs an example in the docstring of how to get reversed order.

File: networkx/algorithms/dag.py (hunk 6 above)
Comment (MisterSheik) [inline networkx/algorithms/dag.py]: Ok.

File: networkx/algorithms/dag.py (hunk 1 above)
Comment (MisterSheik) [inline networkx/algorithms/dag.py]: Because topological_sort is now a generator function, it will not raise that exception until the iterator is consumed.

File: networkx/algorithms/dag.py (hunk 9 above)
Comment (MisterSheik) [inline networkx/algorithms/dag.py]: `reversed(list(topological_sort(…`?

File: networkx/algorithms/dag.py (hunk 3 above)
Comment (jfinkels) [inline networkx/algorithms/dag.py]: Oh I see, well then maybe leave it for a different pull request.

File: networkx/algorithms/dag.py (hunk 9 above)
Comment (jfinkels) [inline networkx/algorithms/dag.py]: Yep, add an examples section:

    Examples
    ---------
    To get the reverse order of the topological sort::

        >>> reversed(list(nx.topological_sort(G)))
        ...

File: networkx/algorithms/dag.py (hunk 2 above)
Comment (MisterSheik) [inline networkx/algorithms/dag.py]: I don't see how to do it with this algorithm, but I think `topological_sort(G.subgraph(nx.descendants(G, nbunch)))` would work for single nodes?

File: networkx/algorithms/dag.py (hunk 8 above)
Comment (MisterSheik) [inline networkx/algorithms/dag.py]: Done.

File: networkx/algorithms/dag.py (hunk 9 above)
Comment (MisterSheik) [inline networkx/algorithms/dag.py]: Done.

**Comment by MisterSheik** [comment]
I've addressed everything in your review @jfinkels .  @hagberg Any chance we can get this graph algorithm checked in soon?

**Comment by hagberg** [comment]
Yes, let's resolve the merge conflicts and then it is OK.
Also we need to document these changes in the 2.0 api doc.

**Comment by MisterSheik** [comment]
@hagberg Sounds good.  I think I should wait for https://github.com/networkx/networkx/pull/1693 since I'm expecting an iterable of pairs from `G.indegree` rather than a dict?

**Comment by hagberg** [comment]
OK - We'll get to #1693 soon.

**Comment by MisterSheik** [comment]
@hagberg Awesome!  I am really looking forward to it.

**Comment by dschult** [comment]
#1662  should be good to go as far as iterable in_degree methods due to commit bfd855f

File: networkx/algorithms/dag.py [hunk 10]
```diff
@@ -79,172 +82,183 @@ def is_directed_acyclic_graph(G):
...
-        order.append(v)
+    while zero_indegree:
+        _, node = heapq.heappop(zero_indegree)
 
-    ancestors = set()
-    explored = set()
-    order = []
+        if node not in G:
+            raise RuntimeError("Graph changed during iteration")
+        for _, child in G.edges(node):
```
Comment (dschult) [inline networkx/algorithms/dag.py]: should this be ```for child in G[node]:```?

Comment (MisterSheik) [inline networkx/algorithms/dag.py]: What is the difference between `G.edges(node)` and `G[node]`?

File: networkx/algorithms/tests/test_dag.py [hunk 11]
```diff
@@ -11,36 +12,15 @@ def setUp(self):
...
-                      nx.topological_sort, DG, reverse=True)
-        assert_raises(nx.NetworkXUnfeasible,
-                      nx.topological_sort_recursive, DG, reverse=True)
+        assert_raises(nx.NetworkXUnfeasible, consume, nx.topological_sort(DG))
 
         DG.remove_edge(2, 3)
-        assert_equal(nx.topological_sort(DG, reverse=True), [2, 3, 1])
-        assert_equal(
-            nx.topological_sort_recursive(DG, reverse=True), [2, 3, 1])
+        consume(nx.topological_sort(DG))
```
Comment (dschult) [inline networkx/algorithms/tests/test_dag.py]: It looks like you no longer test the values returned by the sort... ```consume``` only steps through the iterator, so we are checking whether there are errors during iteration. 

Can we continue to test the values being returned by using ```list``` instead of consume? Or is there something else going on here...?

Comment (MisterSheik) [inline networkx/algorithms/tests/test_dag.py]: The `consume` ensures that there are no exceptions due to cycles.

The old tests worked only by an accident of stable dict iteration in CPython.  They were never guaranteed to work in Python.  They may not work in other implementations of Python (IronPython, PyPy, etc.)  They may not work in newer versions of Python (which use hash randomization) and even newer versions (for which it is proposed to make dict an OrderedDict everywhere).

Since the old tests are essentially broken, I have removed everything but the no-cycle-verification.

Comment (dschult) [inline networkx/algorithms/tests/test_dag.py]: Are you saying that we can't use list(topological_sort(DG)) and get
reproducible results?
I thought the key(node) sorting would remove that issue...

Comment (MisterSheik) [inline networkx/algorithms/tests/test_dag.py]: Right, so `list(lexicographical_topological_sort(DG))` produces reproducible results.  `topological_sort` does not.  What do you propose?

File: networkx/algorithms/dag.py (hunk 10 above)
Comment (dschult) [inline networkx/algorithms/dag.py]: ```G[node]``` is a direct lookup of the successors of ```node```. ```G.edges(node)``` returns outgoing edges that involve ```node```. It wraps the successors into a tuple with the node.  Sometimes the tuple is helpful, but here I think you just unwrap it again to get the successor.  Not a big deal, but G[node] is typically what you want if you already know ```node```.

Comment (hagberg) [inline networkx/algorithms/dag.py]: How about some simple tests?  If there is more than one valid result allow
any.

Comment (MisterSheik) [inline networkx/algorithms/dag.py]: I see, thanks.  I'm glad you pointed it out.  I will make this change now.

File: networkx/algorithms/tests/test_dag.py (hunk 11 above)
Comment (hagberg) [inline networkx/algorithms/tests/test_dag.py]: How about we put the tests back in and check against all possible valid results?  There aren't very many (2?) for the simple tests.

File: networkx/utils/misc.py [hunk 12]
```diff
@@ -175,3 +177,16 @@ def dict_to_numpy_array1(d,mapping=None):
         i = mapping[k1]
         a[i] = d[k1]
     return a
+
+# Recipe from the itertools documentation.
+def consume(iterator):
```
Comment (hagberg) [inline networkx/utils/misc.py]: Do we really need to add `consume`?  And if we do why not the Python recipe with a count?

Comment (MisterSheik) [inline networkx/utils/misc.py]: consume is what I needed in the algorithm.  What's the count recipe?  `sum(1 for _ in iterator)`?

File: networkx/algorithms/tests/test_dag.py (hunk 11 above)
Comment (dschult) [inline networkx/algorithms/tests/test_dag.py]: Can you add the tests back with checking against all possible results? (at least for simple examples)

**Comment by vtslab** [comment]
As @hagberg mentioned 'Also we need to document these changes in the 2.0 api doc.'
This is done, but not in the 2.x release notes API changes https://networkx.github.io/documentation/networkx-2.0/release/release_2.0.html
(I found out the hard way) en also mention this in the 1.x to 2.0 migration guide.
https://networkx.github.io/documentation/networkx-2.0/release/migration_guide_from_1.x_to_2.0.html

**Comment by NeilGirdhar** [comment]
@vtslab Hope this helps!  https://github.com/networkx/networkx/pull/2892
